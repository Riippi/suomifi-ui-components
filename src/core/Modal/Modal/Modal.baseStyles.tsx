import { css } from 'styled-components';
import { defaultThemeTokens as theme } from '../../theme';
import { radius } from '../../theme/radius';
import { alphaHex } from '../../../utils/css';
import { element, font } from '../../theme/reset';

export const baseStyles = css`
  &.fi-modal {
    ${element(theme)}
    ${font(theme)('actionElementInnerTextBold')}
    width: 100vw;
    height: 100vh;
    position: fixed;
    top: 0;
    left: 0;

    &--no-portal {
      z-index: 100;
    }

    & .fi-modal_overlay {
      background-color: ${alphaHex(0.5)(theme.colors.blackBase)};
      width: 100vw;
      height: 100vh;
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center;
    }

    & .fi-modal_content-container {
      border-radius: ${radius.modal};
      background-color: ${theme.colors.whiteBase};
      border-top: ${theme.spacing.insetXs} solid ${theme.colors.highlightBase};
      overflow: hidden;
      max-height: calc(100vh - 50px);
      min-height: 230px;
      width: 800px;
      flex: 0 1 auto;
      display: flex;
      flex-direction: column;
      align-items: stretch;
    }

    &--no-scroll {
      & .fi-modal_content-container {
        width: 540px;
        min-height: 190px;
      }
    }

    &--small-screen {
      & .fi-modal_content-container {
        border-radius: 0;
        width: 100%;
        height: 100%;
        max-height: 100%;
      }
    }
  }
`;
