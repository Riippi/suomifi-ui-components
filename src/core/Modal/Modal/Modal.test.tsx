import React, { ReactNode, useRef, useState } from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import { axeTest } from '../../../utils/test/axe';

import { Modal, ModalProps } from './Modal';
import { ModalContent, ModalFooter, ModalTitle } from '../';
import { Button } from '../../../';

describe('Basic modal', () => {
  const text = 'Modal Content';

  const BasicModal = (props?: Partial<ModalProps>) => (
    <Modal visible={true} usePortal={false} {...props}>
      <ModalContent>
        <ModalTitle>Test modal</ModalTitle>
        <p>{text}</p>
      </ModalContent>
      <ModalFooter>
        <Button>OK</Button>
      </ModalFooter>
    </Modal>
  );

  it('should be hidden when visible is set to false', () => {
    const { container } = render(BasicModal({ visible: false }));
    expect(container.firstChild).toBe(null);
  });

  it('should have given className', () => {
    const { getByRole } = render(BasicModal({ className: 'test-class-name' }));
    expect(getByRole('dialog')).toHaveClass('test-class-name');
  });

  it('should match snapshot', () => {
    const { container } = render(BasicModal());
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not have basic accessibility issues', axeTest(BasicModal()));
});

describe('Modal sibling DOM nodes', () => {
  const ModalWithSiblings = () => {
    const [open, setOpen] = useState(true);
    return (
      <>
        <div data-testid="test-sibling">Test content</div>
        <div
          data-testid="test-sibling-with-aria-hidden-true"
          aria-hidden="true"
          role="presentation"
        >
          Test content with Aria-hidden true
        </div>
        <div
          data-testid="test-sibling-with-aria-hidden-false"
          aria-hidden="false"
          role="presentation"
        >
          Test content with Aria-hidden false
        </div>
        <div
          data-testid="portal-containing-node"
          aria-hidden="false"
          role="button"
        >
          <Modal
            data-testid="modal"
            visible={open}
            usePortal={false}
            onEscKeyDown={() => setOpen(false)}
          >
            <ModalContent>
              <ModalTitle>Test modal</ModalTitle>
              <p>Some test text</p>
            </ModalContent>
            <ModalFooter>
              <Button>OK</Button>
            </ModalFooter>
          </Modal>
          <div data-testid="test-sibling-inside-portal-node">
            Test siblign inside portal node
          </div>
          <div
            data-testid="test-sibling-inside-portal-node-aria-hidden-true"
            aria-hidden="true"
            role="button"
          >
            Test siblign inside portal node aria hidden true
          </div>
          <div
            data-testid="test-sibling-inside-portal-node-aria-hidden-false"
            aria-hidden="false"
            role="button"
          >
            Test siblign inside portal node aria-hidden false
          </div>
        </div>
      </>
    );
  };

  it('containing the modal should not be aria-hidden', () => {
    const { getByTestId, baseElement } = render(<ModalWithSiblings />);
    expect(baseElement).not.toHaveAttribute('aria-hidden');
    expect(getByTestId('portal-containing-node')).toHaveAttribute(
      'aria-hidden',
      'false',
    );
    expect(getByTestId('modal')).not.toHaveAttribute('aria-hidden');
  });

  it('not containing the modal should be aria-hidden', () => {
    const { getByTestId } = render(<ModalWithSiblings />);
    expect(getByTestId('test-sibling')).toHaveAttribute('aria-hidden', 'true');
    expect(getByTestId('test-sibling-with-aria-hidden-true')).toHaveAttribute(
      'aria-hidden',
      'true',
    );
    expect(getByTestId('test-sibling-with-aria-hidden-false')).toHaveAttribute(
      'aria-hidden',
      'true',
    );
    expect(getByTestId('test-sibling-inside-portal-node')).toHaveAttribute(
      'aria-hidden',
      'true',
    );
    expect(
      getByTestId('test-sibling-inside-portal-node-aria-hidden-true'),
    ).toHaveAttribute('aria-hidden', 'true');
    expect(
      getByTestId('test-sibling-inside-portal-node-aria-hidden-false'),
    ).toHaveAttribute('aria-hidden', 'true');
  });

  it('should preserve aria-hidden and role state after closing modal', () => {
    const { getByTestId } = render(<ModalWithSiblings />);
    fireEvent.keyDown(getByTestId('modal'), {
      key: 'Esc',
      code: 27,
      charCode: 27,
    });
    expect(getByTestId('test-sibling')).not.toHaveAttribute('aria-hidden');
    expect(getByTestId('test-sibling-with-aria-hidden-true')).toHaveAttribute(
      'aria-hidden',
      'true',
    );
    expect(getByTestId('test-sibling-with-aria-hidden-true')).toHaveAttribute(
      'role',
      'presentation',
    );
    expect(getByTestId('test-sibling-with-aria-hidden-false')).toHaveAttribute(
      'aria-hidden',
      'false',
    );
    expect(getByTestId('test-sibling-with-aria-hidden-false')).toHaveAttribute(
      'role',
      'presentation',
    );
    expect(getByTestId('portal-containing-node')).toHaveAttribute(
      'aria-hidden',
      'false',
    );
    expect(getByTestId('portal-containing-node')).toHaveAttribute(
      'role',
      'button',
    );
    expect(getByTestId('test-sibling-inside-portal-node')).not.toHaveAttribute(
      'aria-hidden',
    );
    expect(
      getByTestId('test-sibling-inside-portal-node-aria-hidden-true'),
    ).toHaveAttribute('aria-hidden', 'true');
    expect(
      getByTestId('test-sibling-inside-portal-node-aria-hidden-true'),
    ).toHaveAttribute('role', 'button');
    expect(
      getByTestId('test-sibling-inside-portal-node-aria-hidden-false'),
    ).toHaveAttribute('aria-hidden', 'false');
    expect(
      getByTestId('test-sibling-inside-portal-node-aria-hidden-false'),
    ).toHaveAttribute('role', 'button');
  });
});

describe('Modal variant', () => {
  const ModalWithProps = (props?: Partial<ModalProps>) => (
    <Modal visible={true} usePortal={false} {...props}>
      <ModalContent>
        <ModalTitle>Test modal</ModalTitle>
        <p>Modal Content</p>
      </ModalContent>
      <ModalFooter>
        <Button>OK</Button>
        <Button variant="secondary">Cancel</Button>
      </ModalFooter>
    </Modal>
  );

  it('smallScreen should have correct classname', () => {
    const { getByRole } = render(ModalWithProps({ variant: 'smallScreen' }));
    expect(getByRole('dialog')).toHaveClass('fi-modal--small-screen');
  });
});

describe('Modal focus', () => {
  const ModalWithProps = (children: ReactNode, props?: Partial<ModalProps>) => (
    <Modal visible={true} usePortal={false} {...props}>
      <ModalContent>
        <ModalTitle>Test modal</ModalTitle>
        {children}
      </ModalContent>
      <ModalFooter>
        <Button>OK</Button>
        <Button variant="secondary">Cancel</Button>
      </ModalFooter>
    </Modal>
  );

  it('should be on heading by default', async () => {
    const { getAllByRole } = render(ModalWithProps(<p>Modal Content</p>));
    await waitFor(() => expect(getAllByRole('heading')[0]).toHaveFocus());
  });

  it('should be on element provided in props', async () => {
    const RefTest1 = () => {
      const ref = useRef(null);
      return (
        <Modal visible={true} usePortal={false} focusOnOpenRef={ref}>
          <ModalContent>
            <ModalTitle>Test modal</ModalTitle>
            <button>Test button 1</button>
            <button ref={ref}>Test button 2</button>
          </ModalContent>
          <ModalFooter>
            <Button>OK</Button>
            <Button variant="secondary">Cancel</Button>
          </ModalFooter>
        </Modal>
      );
    };
    const { getByText } = render(<RefTest1 />);
    await waitFor(() => expect(getByText('Test button 2')).toHaveFocus());
  });
});

describe('Closing Modal', () => {
  const RefTest = (props?: Partial<ModalProps>) => {
    const [open, setOpen] = useState(false);
    const ref = useRef(null);
    return (
      <>
        <button onClick={() => setOpen(!open)}>Toggle modal</button>
        <input ref={ref} type="text" />
        <Modal
          visible={open}
          usePortal={false}
          focusOnCloseRef={ref}
          {...props}
        >
          <ModalContent>
            <ModalTitle>Test modal</ModalTitle>
            Test Content
            <button onClick={() => setOpen(false)}>Close modal</button>
          </ModalContent>
          <ModalFooter>
            <Button>OK</Button>
            <Button variant="secondary">Cancel</Button>
          </ModalFooter>
        </Modal>
      </>
    );
  };

  it('should be possible with ESC key', () => {
    const mockEsc = jest.fn();
    const { getByText } = render(
      <RefTest focusOnCloseRef={undefined} onEscKeyDown={mockEsc} />,
    );
    const openButton = getByText('Toggle modal');
    fireEvent.click(openButton);
    const content = getByText('Test Content');
    expect(content).toBeTruthy();
    fireEvent.keyDown(content, {
      key: 'Esc',
      code: 27,
      charCode: 27,
    });
    expect(mockEsc).toHaveBeenCalledTimes(1);
  });

  it('should return focus to original element', async () => {
    const { getByText } = render(<RefTest focusOnCloseRef={undefined} />);
    const openButton = getByText('Toggle modal');
    openButton.focus();
    fireEvent.click(openButton);
    const closeButton = getByText('Close modal');
    expect(closeButton).toBeTruthy();
    fireEvent.click(closeButton);
    await waitFor(() => expect(openButton).toHaveFocus());
  });

  it('should return focus to given element', async () => {
    const { getByText, getByRole } = render(<RefTest />);
    const openButton = getByText('Toggle modal');
    fireEvent.click(openButton);
    const closeButton = getByText('Close modal');
    expect(closeButton).toBeTruthy();
    fireEvent.click(closeButton);
    const input = getByRole('textbox');
    await waitFor(() => expect(input).toHaveFocus());
  });
});
