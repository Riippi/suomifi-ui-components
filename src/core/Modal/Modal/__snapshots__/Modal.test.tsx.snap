// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic modal should match snapshot 1`] = `
.c8 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c8:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c8::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c8::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c8::-webkit-inner-spin-button {
  height: auto;
}

.c8::-webkit-outer-spin-button {
  height: auto;
}

.c8::before,
.c8::after {
  box-sizing: border-box;
}

.c1 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c1::before,
.c1::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4::before,
.c4::after {
  box-sizing: border-box;
}

.c0.fi-modal {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  width: 100vw;
  height: 100vh;
  position: fixed;
  top: 0;
  left: 0;
}

.c0.fi-modal--no-portal {
  z-index: 100;
}

.c0.fi-modal .fi-modal_overlay {
  background-color: rgba(41,41,41,0.5);
  width: 100vw;
  height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0.fi-modal .fi-modal_content-container {
  border-radius: 4px;
  background-color: hsl(0,0%,100%);
  border-top: 4px solid hsl(212,63%,45%);
  overflow: hidden;
  max-height: calc(100vh - 50px);
  min-height: 230px;
  width: 800px;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c0.fi-modal--no-scroll .fi-modal_content-container {
  width: 540px;
  min-height: 190px;
}

.c0.fi-modal--small-screen .fi-modal_content-container {
  border-radius: 0;
  width: 100%;
  height: 100%;
  max-height: 100%;
}

.c3.fi-modal_content {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  max-height: 100%;
  padding: 24px 30px 50px 30px;
  overflow-y: auto;
}

.c3.fi-modal_content--no-scroll {
  overflow: hidden;
  padding: 34px 30px 10px 30px;
}

.c3.fi-modal_content--small-screen {
  padding-top: 20px;
  padding-left: 20px;
  padding-bottom: 50px;
  padding-right: 20px;
}

.c5 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  color: hsl(0,0%,16%);
}

.c5.fi-heading--h1hero {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 40px;
  line-height: 48px;
  font-weight: 600;
}

.c5.fi-heading--h1 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 40px;
  line-height: 48px;
  font-weight: 300;
}

.c5.fi-heading--h2 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 28px;
  line-height: 34px;
  font-weight: 600;
}

.c5.fi-heading--h3 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 22px;
  line-height: 28px;
  font-weight: 600;
}

.c5.fi-heading--h4 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 24px;
  font-weight: 600;
}

.c5.fi-heading--h5 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
}

.c5.fi-heading--h6 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
}

.c5.fi-heading--small-screen {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 32px;
  line-height: 38px;
  font-weight: 300;
}

.c5.fi-heading--small-screen.fi-heading--h1hero {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 32px;
  line-height: 38px;
  font-weight: 600;
}

.c5.fi-heading--small-screen.fi-heading--h1 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 32px;
  line-height: 38px;
  font-weight: 300;
}

.c5.fi-heading--small-screen.fi-heading--h2 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 26px;
  line-height: 32px;
  font-weight: 600;
}

.c5.fi-heading--small-screen.fi-heading--h3 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 20px;
  line-height: 26px;
  font-weight: 600;
}

.c5.fi-heading--small-screen.fi-heading--h4 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
}

.c5.fi-heading--small-screen.fi-heading--h5 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
}

.c5.fi-heading--small-screen.fi-heading--h6 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 20px;
  font-weight: 600;
}

.c6.fi-modal_title {
  padding-bottom: 20px;
}

.c6.fi-modal_title--no-scroll {
  padding-bottom: 10px;
}

.c6.fi-modal_title--small-screen {
  padding-bottom: 20px;
}

.c7.fi-modal_footer {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  position: relative;
}

.c7.fi-modal_footer .fi-modal_footer_content {
  padding-top: 0;
  padding-right: 15px;
  padding-bottom: 20px;
  padding-left: 30px;
}

.c7.fi-modal_footer .fi-modal_footer_content > * {
  margin-top: 20px;
  margin-right: 15px;
}

.c7.fi-modal_footer .fi-modal_footer_content-gradient-overlay {
  position: absolute;
  height: 51px;
  width: 100%;
  top: -51px;
  border-bottom: 1px solid hsl(202,7%,80%);
}

.c7.fi-modal_footer .fi-modal_footer_content-gradient-overlay .fi-modal_footer_content-gradient {
  height: 50px;
  margin: 0 20px 0 20px;
  background: linear-gradient( 0deg, hsl(0,0%,100%), rgba(255,255,255,0) );
}

.c7.fi-modal_footer--small-screen .fi-modal_footer_content {
  padding-right: 5px;
  padding-bottom: 15px;
  padding-left: 20px;
}

.c7.fi-modal_footer--small-screen .fi-modal_footer_content > * {
  display: block;
  width: calc(100% - 15px);
  margin-top: 15px;
}

.c9 {
  color: hsl(0,0%,16%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  font-size: 14px;
  line-height: 20px;
  padding: 10px 20px;
  min-height: 40px;
  color: hsl(0,0%,100%);
  background: linear-gradient(0deg,hsl(212,63%,37%) 0%,hsl(212,63%,45%) 100%);
  border-radius: 2px;
  text-align: center;
  text-shadow: 0 1px 1px rgba(0,53,122,0.5);
  cursor: pointer;
}

.c9:focus {
  outline: none;
  position: relative;
}

.c9:focus::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
}

.c9:hover {
  background: linear-gradient(0deg,hsl(212,63%,45%) 0%,hsl(212,63%,49%) 100%);
}

.c9:active {
  background: hsl(212,63%,37%);
}

.c9.fi-button--disabled,
.c9[disabled],
.c9:disabled {
  text-shadow: 0 1px 1px rgba(41,41,41,0.5);
  background: linear-gradient(0deg,hsl(202,7%,67%) 0%,hsl(202,7%,80%) 100%);
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: not-allowed;
}

.c9.fi-button--disabled::after {
  border: none;
  box-shadow: none;
}

.c9.fi-button--fullwidth {
  display: block;
  width: 100%;
}

.c9.fi-button--inverted {
  background: none;
  background-color: hsl(212,63%,45%);
  border: 1px solid hsl(0,0%,100%);
  text-shadow: none;
}

.c9.fi-button--inverted:hover {
  background: linear-gradient(-180deg,rgba(255,255,255,0.1) 0%,rgba(255,255,255,0) 100%);
}

.c9.fi-button--inverted:active {
  background: none;
  background-color: hsl(212,63%,45%);
}

.c9.fi-button--inverted.fi-button--disabled,
.c9.fi-button--inverted[disabled],
.c9.fi-button--inverted:disabled {
  opacity: 0.41;
  background: none;
  background-color: none;
}

.c9.fi-button--secondary {
  color: hsl(212,63%,45%);
  background: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(212,63%,45%);
  text-shadow: none;
}

.c9.fi-button--secondary:hover {
  background: linear-gradient(-180deg,hsl(202,7%,93%) 0%,rgba(255,255,255,0) 100%);
}

.c9.fi-button--secondary:active {
  background: none;
  background-color: hsl(202,7%,93%);
}

.c9.fi-button--secondary.fi-button--disabled,
.c9.fi-button--secondary[disabled],
.c9.fi-button--secondary:disabled {
  color: hsl(202,7%,67%);
  text-shadow: none;
  background: none;
  background-color: hsl(212,63%,98%);
  border-color: hsl(202,7%,67%);
}

.c9.fi-button--secondary-noborder {
  color: hsl(212,63%,45%);
  background: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(212,63%,45%);
  text-shadow: none;
  border: none;
  background-color: transparent;
}

.c9.fi-button--secondary-noborder:hover {
  background: linear-gradient(-180deg,hsl(202,7%,93%) 0%,rgba(255,255,255,0) 100%);
}

.c9.fi-button--secondary-noborder:active {
  background: none;
  background-color: hsl(202,7%,93%);
}

.c9.fi-button--secondary-noborder.fi-button--disabled,
.c9.fi-button--secondary-noborder[disabled],
.c9.fi-button--secondary-noborder:disabled {
  color: hsl(202,7%,67%);
  text-shadow: none;
  background: none;
  background-color: hsl(212,63%,98%);
  border-color: hsl(202,7%,67%);
}

.c9.fi-button--link {
  color: hsl(212,63%,45%);
  color: hsl(212,63%,45%);
  background: none;
  background-color: hsl(0,0%,100%);
  border: 1px solid hsl(212,63%,45%);
  text-shadow: none;
  background: linear-gradient(0deg,hsl(215,100%,95%),hsl(215,100%,97%));
  border: none;
}

.c9.fi-button--link:hover {
  background: linear-gradient(-180deg,hsl(202,7%,93%) 0%,rgba(255,255,255,0) 100%);
}

.c9.fi-button--link:active {
  background: none;
  background-color: hsl(202,7%,93%);
}

.c9.fi-button--link.fi-button--disabled,
.c9.fi-button--link[disabled],
.c9.fi-button--link:disabled {
  color: hsl(202,7%,67%);
  text-shadow: none;
  background: none;
  background-color: hsl(212,63%,98%);
  border-color: hsl(202,7%,67%);
}

.c9.fi-button--link:hover {
  background: linear-gradient(0deg,hsl(215,100%,97%),hsl(212,63%,98%));
}

.c9.fi-button--link:active {
  background: linear-gradient(0deg,hsl(202,7%,93%),hsl(202,7%,97%));
}

.c9.fi-button--link.fi-button--disabled,
.c9.fi-button--link[disabled],
.c9.fi-button--link:disabled {
  color: hsl(202,7%,67%);
  background: none;
  background-color: hsl(202,7%,97%);
}

.c9 > .fi-button_icon {
  width: 16px;
  height: 16px;
  margin-right: 8px;
  vertical-align: middle;
  -webkit-transform: translateY(-0.1em);
  -ms-transform: translateY(-0.1em);
  transform: translateY(-0.1em);
}

.c9 > .fi-button_icon.fi-button_icon--right {
  margin-right: 0;
  margin-left: 8px;
}

.c9.fi-button--disabled > .fi-button_icon {
  cursor: not-allowed;
}

<div
  aria-describedby="3_title"
  aria-modal="true"
  class="c0 fi-modal fi-modal--no-portal"
  role="dialog"
>
  <div
    class="c1 fi-modal_overlay"
  >
    <div
      class="c2 fi-modal_content-container"
    >
      <div
        class="c1 fi-modal_content c3"
      >
        <h2
          class="c4 fi-heading c5 c6 fi-modal_title fi-heading--h3"
          id="3_title"
        >
          Test modal
        </h2>
        <p>
          Modal Content
        </p>
      </div>
      <div
        class="c1 c7 fi-modal_footer"
      >
        <div
          class="c1 fi-modal_footer_content"
        >
          <button
            aria-disabled="false"
            class="c8 fi-button c9"
            tabindex="0"
            type="button"
          >
            OK
          </button>
        </div>
        <div
          class="c1 fi-modal_footer_content-gradient-overlay"
        >
          <div
            class="c1 fi-modal_footer_content-gradient"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
